### PR TITLE
[WiP] [JENKINS-26462] Extract maven builder to a separate plugin

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-26462](https://issues.jenkins-ci.org/browse/JENKINS-26462)

Not ready to merge yet. Extracting to: https://github.com/Vlatombe/maven-builder-plugin

### CODE REVIEW

**Comment by jglick** [comment]
Cf. https://github.com/kohsuke/git-filter-branch-tools.

**Comment by daniel-beck** [comment]
@Vlatombe Are you still interested in this PR?

**Comment by jglick** [comment]
For splitting plugins, a PR which _deletes_ the plugin sources is not so useful, since (a) it will have tons of merge conflicts over time, (b) it is hard to review what is being moved. Anyway when you merge it, you will no longer have the plugin sources handy.

Might be better to create a PR which merely _moves_ the sources to be split into a new Maven module with `jenkins-module` packaging, included in `war/pom.xml`. Thus the PR contains only renames and changes to any other sources you needed to fix; and merging it does not really change runtime behavior, as the code is still in the core class loader. Then you do the actual split—deleting the module from `jenkinsci/jenkins` sources, importing its history into a new plugin repository, releasing that plugin, listing it in `DETACHED_PLUGINS`—with direct commits, not PRs. (Due to the nature of Maven releases, you cannot really prepare such changes offline.)

I would have to look up some examples of splits I have done to remember exactly what works best.

**Comment by jglick** [comment]
(You can find resolved JIRA issues with label `split-plugins-from-core` and follow the history.)

**Comment by jglick** [comment]
One problem that I know comes up is this: say you do the split during `2.3-SNAPSHOT` of core. So you want the plugin to depend on core 2.3, so that 2.2 users will not see it as a valid installation option, since then they would have two copies. But you cannot release 2.3 until the detached list mentions version 1.0 of the plugin. And Maven will not let you release 1.0 with a snapshot dependency! So it is a bit of a Catch-22. The workaround is to release plugin 1.0 with a 2.2 dependency, and then as soon as 2.3 is released, update the plugin to depend on it, releasing plugin 1.1.

**Comment by alvarolobato** [comment]
@jglick thanks for the comments, having a look at old splits on jira. Already have a list of files to move and going to start the first step to move inside jenkins project.

File: core/src/main/java/hudson/model/Project.java [hunk 1]
```diff
@@ -60,7 +57,7 @@
  * @author Kohsuke Kawaguchi
  */
 public abstract class Project<P extends Project<P,B>,B extends Build<P,B>>
-    extends AbstractProject<P,B> implements SCMTriggerItem, Saveable, ProjectWithMaven, BuildableItemWithBuildWrappers {
```
Comment (alvarolobato) [inline core/src/main/java/hudson/model/Project.java]: @Vlatombe @jglick  won't this change impact how other plugins? I've done a search on github and many plugins including maven plugin check for this interface.

@Vlatombe did you check the consequences when you prepared this PR?

Comment (Vlatombe) [inline core/src/main/java/hudson/model/Project.java]: Argh, I didn't see this, but it is gonna be a PITA.

Comment (alvarolobato) [inline core/src/main/java/hudson/model/Project.java]: @jglick can I have your opinion on how to proceed with this dependency?. Thanks

**Comment by daniel-beck** [comment]
@Vlatombe Can we close this in favor of #2352?

**Comment by Vlatombe** [comment]
Sure

**Pull request closed without merging**